text_classification.py

```diff
@@ -0,0 +1,487 @@
+import os
+import re
+import torch
+from torchtext.utils import download_from_url, extract_archive
+from torchtext.data.utils import generate_ngrams
+from torchtext.vocab import build_vocab
+import random
+from torchtext import data
+from tqdm import tqdm
+from torchtext.data import dataset
+from torchtext.data.iterator import generate_iterators
+
+
+class TextDataset(torch.utils.data.Dataset):
+    def __init__(self, examples, fields):
+        self.examples = examples
+        self.fields = fields
+
+    def __getitem__(self, i):
+        return self.examples[i]
+
+    def __len__(self):
+        try:
+            return len(self.examples)
+        except TypeError:
+            return 2**32
+
+    def __iter__(self):
+        for x in self.examples:
+            yield x
+
+
+class TextClassificationDataset(torch.utils.data.Dataset):
+    """Defines text classification datasets.
+        Currently, we only support the following datasets:
+
+             - AG_NEWS
+             - SogouNews
+             - DBpedia
+             - YelpReviewPolarity
+             - YelpReviewFull
+             - YahooAnswers
+             - AmazonReviewPolarity
+             - AmazonReviewFull
+
+    """
+
+    def __init__(self, dataset_name, root='.data',
+                 text_field=None, label_field=None, ngrams=1):
+        """Create a text classification dataset instance.
+
+        Arguments:
+            dataset_name: The name of dataset, include "ag_news", "sogou_news",
+                "dbpedia", "yelp_review_polarity", "yelp_review_full",
+                "yahoo_answers", "amazon_review_full", "amazon_review_polarity".
+            root: Directory where the dataset are saved. Default: ".data"
+            text_field: The field that will be used for the sentence. If not given,
+                'spacy' token will be used.
+            label_field: The field that will be used for the label. If not given,
+                'float' token will be used.
+            ngrams: a contiguous sequence of n items from s string text.
+                Default: 1
+
+        Examples:
+            >>> txt_cls = TextClassificationDataset("ag_news", ngrams=2)
+
+        """
+        self.dataset_name = dataset_name
+        if dataset_name in supported_text_classification_datasets.keys():
```

Comment:
You have the class instances below (e.g. AG_NEWS), so you use this to deduplicate some code I think.
 The main reason for that seems to be that you're using "load_text_classification_data" and have a predefined split.

Reply:
This interface is old, and not generic. The major function components in this class should be "download", "preprocess", "load token list", "general iterator/batch". That's all.